src/main/kotlin/net/ccbluex/liquidbounce/script/bindings/api/ScriptAsyncUtil.kt

```diff
@@ -0,0 +1,196 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with LiquidBounce. If not, see <https://www.gnu.org/licenses/>.
+ */
+package net.ccbluex.liquidbounce.script.bindings.api
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.launch
+import net.ccbluex.liquidbounce.api.core.HttpClient
+import net.ccbluex.liquidbounce.event.EventListener
+import net.ccbluex.liquidbounce.event.events.GameTickEvent
+import net.ccbluex.liquidbounce.event.handler
+import net.ccbluex.liquidbounce.script.ScriptApiRequired
+import net.ccbluex.liquidbounce.utils.client.mc
+import net.ccbluex.liquidbounce.utils.kotlin.EventPriorityConvention.FIRST_PRIORITY
+import org.graalvm.polyglot.Value
+import org.graalvm.polyglot.proxy.ProxyExecutable
+import java.util.function.BooleanSupplier
+import java.util.function.Consumer
+import java.util.function.Supplier
+
+/**
+ * Async utils including game-based tick scheduling and network requests.
+ * JavaScript-only.
+ *
+ * @author MukjepScarlet
+ */
+class ScriptAsyncUtil(
+    private val jsPromiseConstructor: Value
+) {
+
+    companion object TickScheduler : EventListener {
+
+        /** Client HTTP requests */
+        private val scriptJobHolder = SupervisorJob()
+        private val scriptIoScope = CoroutineScope(Dispatchers.IO + scriptJobHolder)
+
+        private val currentTickTasks = arrayListOf<BooleanSupplier>()
+        private val nextTickTasks = arrayListOf<BooleanSupplier>()
+
+        @Suppress("unused")
+        private val tickHandler = handler<GameTickEvent>(priority = FIRST_PRIORITY) {
+            currentTickTasks.removeIf { it.asBoolean }
+            currentTickTasks += nextTickTasks
+            nextTickTasks.clear()
+        }
+
+        private fun schedule(breakLoop: BooleanSupplier) {
+            mc.execute { nextTickTasks += breakLoop }
+        }
+
+        fun clear() {
+            mc.execute {
+                currentTickTasks.clear()
+                nextTickTasks.clear()
+                scriptJobHolder.cancelChildren()
+            }
+        }
+    }
+
+    /**
+     * `Promise.resolve(0)`
+     */
+    private val defaultPromise: Value =
+        jsPromiseConstructor.invokeMember("resolve", 0)
+
+    /**
+     * Example: `await ticks(10)`
+     *
+     * @return `Promise<number>`
+     */
+    @ScriptApiRequired
+    fun ticks(n: Int): Value {
+        if (n == 0) {
+            return defaultPromise
+        }
+
+        var remains = n
+        return until { --remains == 0 }
+    }
+
+    /**
+     * Example: `await seconds(1)`
+     *
+     * @return `Promise<number>`
+     */
+    @ScriptApiRequired
+    fun seconds(n: Int): Value = ticks(n * 20)
```

Comment:
I don't think this method should exist. Seconds are a confusing concept in the context of Minecraft (especially since one real-time second does not necessarily equal 20 in game ticks). The `ticks` method alone should be sufficient.